pyro/infer/tmc.py

```diff
@@ -0,0 +1,195 @@
+    log_probs = []
+    for role, trace in zip(("model", "guide"), (model_trace, guide_trace)):
+        for name, site in trace.nodes.items():
+            if site["type"] != "sample" or site["is_observed"]:
+                continue
+            if role == "model" and name in guide_trace:
+                continue
+
+            log_prob, log_denom = compute_site_dice_factor(site)
+            if not is_identically_zero(log_denom):
+                log_prob = log_prob - log_denom
+            if not is_identically_zero(log_prob):
+                log_probs.append(log_prob)
+
+    return log_probs
+
+
+def _compute_tmc_factors(model_trace, guide_trace):
+    # factors
+    log_factors = []
+    for name, site in guide_trace.nodes.items():
+        if site["type"] != "sample" or site["is_observed"]:
+            continue
+        log_factors.append(packed.neg(site["packed"]["log_prob"]))
+    for name, site in model_trace.nodes.items():
+        if site["type"] != "sample":
+            continue
+        if site["name"] not in guide_trace and \
+                not site["is_observed"] and \
+                site["infer"].get("enumerate", None) == "parallel" and \
+                site["infer"].get("num_samples", -1) > 0:
+            # site was sampled from the prior, proposal term cancels log_prob
+            continue
+        log_factors.append(site["packed"]["log_prob"])
+    return log_factors
+
+
+def _compute_tmc_estimate(model_trace, guide_trace):
+
+    # factors
+    log_factors = _compute_tmc_factors(model_trace, guide_trace)
+    log_factors += _compute_dice_factors(model_trace, guide_trace)
+
+    if not log_factors:
+        return 0.
+
+    # loss
+    eqn = ",".join([f._pyro_dims for f in log_factors]) + "->"
+    plates = "".join(frozenset().union(list(model_trace.plate_to_symbol.values()),
+                                       list(guide_trace.plate_to_symbol.values())))
+    tmc, = einsum(eqn, *log_factors, plates=plates,
+                  backend="pyro.ops.einsum.torch_log",
+                  modulo_total=False)
+    return tmc
+
+
+class TensorMonteCarlo(ELBO):
+    """
+    A trace-based implementation of Tensor Monte Carlo [1]
+    by way of Tensor Variable Elimination [2] that supports:
+    - local parallel sampling over any sample site in the model or guide
+    - exhaustive enumeration over any sample site in the model or guide
+
+    To take multiple samples, mark the site with
+    ``infer={'enumerate': 'parallel', 'num_samples': N}``.
+    To configure all sites in a model or guide at once,
+    use :func:`~pyro.infer.enum.config_enumerate`.
+    To enumerate or sample a sample site in the ``model``,
+    mark the site and ensure the site does not appear in the ``guide``.
+
+    This assumes restricted dependency structure on the model and guide:
+    variables outside of an :class:`~pyro.plate` can never depend on
+    variables inside that :class:`~pyro.plate`.
```

Comment:
Note I sometimes get inscrutable sphinx errors with "`" followed by "." in sphinx links. The workaround is to always leave a space before the ".".